RE: The /shelfmodule/add/ page throws a 500 error if the Endpoint is not specified

xlisachan opened the pull request:
Fixes #15282 

This pull requests fixes the error when the endpoint is not specified.

Please review when you have a moment. Thanks!

eviljeff commented on src/olympia/shelves/forms.py:
```diff
@@ -15,14 +15,19 @@ class Meta:
+            return self.errors
+
+        criteria = cleaned_data.get('criteria')
+        endpoint = cleaned_data.get('endpoint')
```
these couple of lines that replace `data['endpoint']` with `data.get('endpoint')` (and same with `criteria`) are the fix for the issue, aren't they?  I don't understand what the rest of the change in the patch is for.

xlisachan replied:
There would still be a `KeyError` if any of the required fields were left blank when attempting to save the form. The fix of the issue was to return any errors first with:
`if any(self.errors):`
 `            return self.errors`

I guess I didn't necessarily have to replace `data['endpoint']` with `data.get('endpoint')`, but I read that was a better way to write it. The rest of the changes were just to move the error message above the field, as opposed to the top of the form, which is more of a preference and not necessary for the the patch. Please let me know if you want me to revert those revisions.

eviljeff commented:
You don't need all of this code and custom handling for errors.  It just needs to only try to ping the endpoint if endpoint is specified.  That endpoint is _required_ is already handled by django forms code and will raise appropriate error messages.

```diff
------------------------- src/olympia/shelves/forms.py -------------------------
index 4d3d9a4ae0..6da6d5d41a 100644
@@ -16,18 +16,21 @@ class ShelfForm(forms.ModelForm):
 
     def clean(self):
         data = self.cleaned_data
-        criteria = data['criteria']
+        criteria = data.get('criteria')
+        endpoint = data.get('endpoint')
         baseUrl = settings.INTERNAL_SITE_URL
 
-        if data['endpoint'] == 'search':
+        if endpoint == 'search':
             api = drf_reverse('v4:addon-search')
             url = baseUrl + api + criteria
-        elif data['endpoint'] == 'collections':
+        elif endpoint == 'collections':
             api = drf_reverse('v4:collection-addon-list', kwargs={
                 'user_pk': settings.TASK_USER_ID,
                 'collection_slug': criteria
             })
             url = baseUrl + api
+        else:
+            return
 
         try:
             response = requests.get(url)
```
seems to fix the issue.

eviljeff commented on src/olympia/shelves/tests/test_forms.py:
```diff
@@ -17,44 +14,39 @@ def setUp(self):
             responses.GET,
-            baseUrl + drf_reverse('v4:addon-search') +
-            self.criteria_sea,
+            reverse_ns('addon-search') + self.criteria_sea,
```
We're specifically pinging the v4 api in the forms code above, because that's the version of the api that addons-frontend is currently using.  If you want to use `reverse_ns`, specify `api_version`.

eviljeff commented on src/olympia/shelves/tests/test_forms.py:
```diff
@@ -17,44 +14,39 @@ def setUp(self):
-                                  'collection_slug': self.criteria_col}),
+            reverse_ns('collection-addon-list', kwargs={
+                'user_pk': settings.TASK_USER_ID,
+                'collection_slug': self.criteria_col}),
```
same here with `api_version`

eviljeff replied:
and etc in this file.

eviljeff commented on src/olympia/shelves/tests/test_forms.py:
```diff
@@ -81,44 +73,40 @@ def test_clean_search_returns_404(self):
-            form.clean()
-        assert exc.exception.message == (
-            'Check criteria - No data found')
+        form.is_valid()
```
so, now the form _is_ valid?  Are there tests that prove that these changes still result in an error here prevent the Shelf from being saved?

xlisachan replied:
I removed the unnecessary changes to the form/tests, and added a test to confirm that `def clean` checks if there are missing required fields.

xlisachan commented:
This latest push includes the following updates:
- Removed the unneccessary changes
- Added the `api_version` to the tests
- Added a test to confirm that `def clean` checks required fields

Please review when you have a moment. Thank you!
